src/main/clojure/clara/tools/fact_graph.clj

```diff
@@ -0,0 +1,90 @@
+;; these nodes are not equal to each other.  This ensures that the number of the internal
+;; nodes accurately reflects the cardinality of each fact in the session.
+(def ^:private ^:dynamic *id-counter* nil)
+
+(defn ^:private add-edge [graph a b]
+  (-> graph
+      (update-in [:forward-edges a] (fnil conj #{}) b)
+      (update-in [:backward-edges b] (fnil conj #{}) a)))
+
+(defn ^:private add-insertion-to-graph
+  [original-graph fact-inserted {:keys [rule explanation]}]
+  (let [facts-direct (sequence
+                      (comp (remove :facts-accumulated)
+                            (map :fact))
+                      (:matches explanation))
+        
+        activation-node (map->RuleActivationNode {:rule rule
+                                                  :id (swap! *id-counter* inc)})
+
+        accum-matches (filter :facts-accumulated (:matches explanation))]
+    
+    (as-> original-graph graph
+      (if (seq accum-matches)
+        (reduce (fn [reduce-graph accum-match]
+                  (let [accum-node (->AccumulationNode (swap! *id-counter* inc))
+                        accum-result (->AccumulationResultNode (swap! *id-counter* inc) (:fact accum-match))]
+                    (as-> reduce-graph g
+                      ;; Add edges from individual facts to an AccumulationResultNode.
+                      (reduce (fn [g accum-element]
+                                (add-edge g accum-element accum-node))
+                              g (:facts-accumulated accum-match))
+                      (add-edge g accum-node accum-result)
+                      (add-edge g accum-result activation-node))))
+                graph
+                accum-matches)
+        graph)
+      ;; Add edges to the rule activation node from the facts that contributed
+      ;; to the rule firing that were not in accumulator condition.
+      (reduce (fn [g f]
+                (add-edge g f activation-node))
+              graph
+              facts-direct)
+      (add-edge graph activation-node fact-inserted))))
+
+(defn session->fact-graph
+  "Given a session, return a graph structure connecting all facts to the facts
+   that they caused to be logically inserted.  Note that such connections will not
+   be made for unconditionally inserted facts."
+  [session]
+  (binding [*id-counter* (atom 0)]
+    (let [fact->explanations (@#'i/gen-fact->explanations session)
+          insertion-tuples (into []
+                                 (comp
+                                  (map (fn [[k v]]
+                                         (map (fn [v-entry]
+                                                [k (update v-entry
+                                                           :rule
+                                                           :name)])
+                                              v)))
+                                  cat)
+                                 fact->explanations)]
```

Comment:
Not a blocker, but a little explanation of this composed function with nested maps wouldn't hurt, or perhaps more descriptive variable names in the nested functions. I had to read it a couple times to untangle the logic here.